feat!: Improves error handling.

amanda-tarafa opened the pull request:
In particular includes more information on GoogleApiException.Message when possible.

Closes #8206

BREAKING CHANGE: If calling code depended on exact exception message.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryInsertResults.cs:
```diff
@@ -135,14 +136,28 @@ public BigQueryInsertResults ThrowOnAnyError()
             }
 
-            var exception = new GoogleApiException(_client.Service.Name, $"Error inserting data. Status: {Status}")
+            // If we attempted to insert just one row, let's be more explicit about the error message.
```
Any reason this is only if we're inserting a single row? Is that because if we've got a lot of errors, we don't want the ToString representation to be huge? It feels a little inconsistent that client code can't use the Error property if they're inserting more than one row... but we may need to accept that as a practical trade-off.

jskeet replied:
(Perhaps we could create a GoogleApiException containing the first error unconditionally, along with the number of errors, and make the message explain that the caller should inspect the BigQueryInsertResult for more information?)

amanda-tarafa replied:
Client code can always use the Error property, the only difference here is what the Error.Message is (which will be part of the GoogleApiException.Message). If we inserted only one row, we can use one error message for that row as the exception message, but if we inserted many and many failed then the exception message could be huge.

I can actually change this to include more cases, for instance, even if we inserted many rows, if just one row failed, etc.

I'll also make the code clearer so that is very explicit that the only thing different is the Error.Message.

jskeet replied:
Right - and the one time that client code *can't* use the error property is because they've called InsertRow which automatically throws, and in that case they'll have the error here.

I think unconditionally including the first error but indicating how many total errors there are could be really useful.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs:
```diff
@@ -107,17 +107,19 @@ public BigQueryJob ThrowOnAnyError()
+                            Message = error.Message
+                        }).ToList(),
+                        Message = errors.Count == 1
+                            ? errors.First().Message
```
If there's exactly one error, do we still want the job reference to be easily available? (Just trying to make sure we don't make information harder to use by trying to simplify things.)

amanda-tarafa replied:
Yes, I'll reformat this message to include the job reference as well.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs:
```diff
@@ -107,17 +107,19 @@ public BigQueryJob ThrowOnAnyError()
+                        }).ToList(),
+                        Message = errors.Count == 1
+                            ? errors.First().Message
+                            : $"Job {Reference?.ProjectId}/{Reference?.Location}/{Reference?.JobId} contained errors"
```
Would it be worth including the number of errors here?

amanda-tarafa replied:
Yep, I'll do that.

jskeet reviewed (commented):
Various questions, but once we've hammered out the details I think we'll have a much nicer experience.

amanda-tarafa reviewed (commented):
Jon, I've replied to your questions, PTAL.

amanda-tarafa reviewed (commented):
Jon, all comments addressed, PTAL. Thanks!

amanda-tarafa commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs:
```diff
@@ -107,17 +107,19 @@ public BigQueryJob ThrowOnAnyError()
+                        }).ToList(),
+                        Message = $"Job {Reference?.ProjectId}/{Reference?.Location}/{Reference?.JobId} contained errors." +
+                        $"Message: {errors.First().Message}" +
+                        (errors.Count == 1 ? "" : $" And {errors.Count - 1} more errors.")
```
I had the period in and removed it because I saw on the test that then we got two periods. But obviously it probably depends, some might come without.

I'll follow your proposal.

amanda-tarafa commented:
Jon, second commit with the changes in order of error count and first message.

Also, I added the `do not merge` label because we need to wait for a release of Apiary BQ, but still this PR won't build until we have updated the dependencies.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs:
```diff
@@ -107,17 +107,18 @@ public BigQueryJob ThrowOnAnyError()
+                            Reason = error.Reason,
+                            Message = error.Message
+                        }).ToList(),
+                        Message = $"Job {Reference?.ProjectId}/{Reference?.Location}/{Reference?.JobId} contained {errors.Count} error(s)." +
```
Nit: add a space after the period at the end of this line.

jskeet reviewed (approved):
Definitely better - one nit about a space between "error(s)." and "First error message" but that's all.

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryInsertResults.cs:
```diff
@@ -135,14 +136,17 @@ public BigQueryInsertResults ThrowOnAnyError()
-                    Errors = Errors.SelectMany(rowErrors => rowErrors).ToList()
+                    Errors = flattenedErrors,
+                    Message = $"Error inserting data: {flattenedErrors.Count} error(s). " +
+                    $"Status: {Status}. " +
```
Nit: indent these two lines (as they're part of Message)

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryInsertResults.cs:
```diff
@@ -135,14 +136,17 @@ public BigQueryInsertResults ThrowOnAnyError()
                 return this;
```
Please could you fix the typo of "excetpion" on line 129 while you're here?

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryJob.cs:
```diff
@@ -107,17 +107,18 @@ public BigQueryJob ThrowOnAnyError()
+                            Message = error.Message
+                        }).ToList(),
+                        Message = $"Job {Reference?.ProjectId}/{Reference?.Location}/{Reference?.JobId} contained {errors.Count} error(s). " +
+                        $"First error message: {errors.First().Message}"
```
More indentation.

amanda-tarafa replied:
Done

jskeet commented on apis/Google.Cloud.BigQuery.V2/Google.Cloud.BigQuery.V2/BigQueryInsertResults.cs:
```diff
@@ -135,14 +136,17 @@ public BigQueryInsertResults ThrowOnAnyError()
             }
 
-            var exception = new GoogleApiException(_client.Service.Name, $"Error inserting data. Status: {Status}")
+            var flattenedErrors = Errors.SelectMany(rowErrors => rowErrors).ToList();
```
At some point we should change Errors to an automatically-implemented property and remove _errors, so that we don't access them inconsistently. Probably best to do that in a separate PR though. (And I can do it.)

amanda-tarafa replied:
Acked, happy to do that when I'm back

jskeet reviewed (approved):
A couple of formatting nits, but basically fine.

amanda-tarafa reviewed (commented):
All nits addressed, I'll merge on green. Thanks!
